python/ray/tune/examples/xgboost_dynamic_resources_example.py

```diff
@@ -0,0 +1,237 @@
+    base_scheduler = ASHAScheduler(
+        max_t=16,  # 16 training iterations
+        grace_period=1,
+        reduction_factor=2)
+
+    def resource_allocation_function(trial_runner: "trial_runner.TrialRunner",
+                                     trial: Trial, result: dict,
+                                     base_trial_resource: PlacementGroupFactory
+                                     ) -> Union[None, PlacementGroupFactory]:
+        """This is a basic example of a resource allocating function.
+
+        The function naively balances free resources (CPUs and GPUs) between
+        trials, giving them all equal priority, ensuring that all resources
+        are always being used. If for some reason a trial ends up with
+        more resources than there are free ones, it will adjust downwards.
+
+        It will also ensure that trial as at least as many resources as
+        it started with (base_trial_resource).
+
+        This function returns a new PlacementGroupFactory with updated
+        resource requirements, or None. If the returned PlacementGroupFactory
+        is equal by value to the one the trial has currently, the scheduler
+        will skip the update process internally (same with None)."""
+
+        if result["training_iteration"] < 1:
+            return None
+
+        min_cpu = base_trial_resource.required_resources.get("CPU", 0)
+        min_gpu = base_trial_resource.required_resources.get("GPU", 0)
+
+        total_available_cpus = (
+            trial_runner.trial_executor._avail_resources.cpu)
+        total_available_gpus = (
+            trial_runner.trial_executor._avail_resources.gpu)
+
+        if min_cpu == 0:
+            upper_cpu_limit = 0
+        else:
+            upper_cpu_limit = math.ceil(total_available_cpus / len(
+                trial_runner.get_live_trials()) / min_cpu)
+
+        if min_gpu == 0:
+            upper_gpu_limit = 0
+        else:
+            upper_gpu_limit = math.ceil(total_available_gpus / len(
+                trial_runner.get_live_trials()) / min_gpu)
+
+        def get_used_cpus_and_gpus(t: Trial):
+            return (t.placement_group_factory.required_resources.get("CPU", 0),
+                    t.placement_group_factory.required_resources.get("GPU", 0))
+
+        trial_used_cpus, trial_used_gpus = get_used_cpus_and_gpus(trial)
+
+        used_cpus_and_gpus = [
+            get_used_cpus_and_gpus(t) for t in trial_runner.get_live_trials()
+        ]
+        used_cpus, used_gpus = zip(*used_cpus_and_gpus)
+        used_cpus = sum(used_cpus)
+        used_gpus = sum(used_gpus)
+
+        free_cpus = total_available_cpus - used_cpus
+        free_gpus = total_available_gpus - used_gpus
+
+        new_cpu = min(upper_cpu_limit, max(trial_used_cpus + free_cpus,
+                                           min_cpu))
+        new_gpu = min(upper_gpu_limit, max(trial_used_gpus + free_gpus,
+                                           min_gpu))
```

Comment:
This looks good, but two questions:

1. Since this is an example, should we add more comments so people understand what exactly is being done here?

2. This example is not "simple", but quite extensive. A simple example would maybe have 3 or 4 lines of code (you could do something very simple like just dividing the total number of CPUs by the number of running trials). 

So first, I think we could probably move this specific function into some kind of collection/utility, because it would be applicable to different kinds of trainables. Users can then just set `resource_allocation_function = even_cpu_gpu_distribution` or so if they want to use it.

Second, for the simple example, consider replacing it by a CPU-only, very specific allocation function?

Reply:
Yeah, I think you are right. How about making this the default resource allocation function, and the example being much simpler? I agree it needs more comments regardless.